### PR TITLE
feat(orchestration): lookup `ChainInfo` from address value

### LINKED ISSUE #10349: orchestration: look up chainId from chain address value
## What is the Problem Being Solved?

As a FastUSDC developer, I need to determine a `chainId` from a `ChainAddress.value` in order to send an IBC transfer.

See Codecider for more context: https://codecider-app.vercel.app/d/a04013d8-1cf8-4c94-a4de-f634419cb199#?!-how-does-the-contract-figure-out-the-chainid-of-the-dydx123-receiver

## Description of the Design

- add `bech32Prefix` to `CosmosChainInfo`
- add a ChainHub method called `getChainInfoByAddresss` that leverages a `bech32Prefix:chainName` mapping to return `ChainInfo`

## Scaling Considerations

`bech32Prefix` must be unique across chains in `ChainHub`. In practice, this is fine for cosmos chains. For EVM chains, where the prefix is always `0x`, a different approach will need to be taken to determine `chainId`.

## Upgrade Considerations

Library code, part of an NPM orch release.

### PR DESCRIPTION
**Pull request by 0xpatrickdev**
closes: #10349

## Description
- add `bech32Prefix?: string` to `CosmosChainInfo` and populate data with `yarn codegen`
- add `getChainInfoByAddress` lookup method to `ChainHub`

### Security Considerations

If multiple cosmos chains use the same `bech32Prefix`, this approach is not sound. In practice, this doesn't seem to be an issue. Developers also control their own ChainHubs and choose whether they want to populate it with data from `agoricNames`. The potential impact is loss of funds from a misrouted transfer.

### Scaling Considerations
- Adds another exo to keep track of `bech32Prefix:chainName` mapping.
- Adds more data to agoricNames (negligible)

### Documentation Considerations
Typedoc will include these changes

### Testing Considerations
Includes unit tests and updates snapshot tests

### Upgrade Considerations
Library code, part of an NPM orch release.

### CODE REVIEW

File: packages/orchestration/src/fetched-chain-info.js [hunk 1]
```diff
@@ -146,36 +148,36 @@ export default /** @type {const} } */ ({
...
           counterPartyPortId: 'transfer',
           ordering: 0,
           state: 3,
           version: 'ics20-1',
         },
       },
-      'osmosis-1': {
-        id: 'connection-2',
-        client_id: '07-tendermint-10',
+      'neutron-1': {
```
Comment (0xpatrickdev) [inline packages/orchestration/src/fetched-chain-info.js]: `yarn codegen` shouldn't create a diff like this (reordering of connection entries).

Need to audit the logic in `convertChainInfo`: https://github.com/Agoric/agoric-sdk/blob/7ae1f278fa8cbeb0cfc777b7cebf507b1f07c958/packages/orchestration/src/utils/registry.js#L120-L121

Comment (0xpatrickdev) [reply packages/orchestration/src/fetched-chain-info.js]: See [832d26f](https://github.com/Agoric/agoric-sdk/pull/10354/commits/832d26f79ad4f8460a71ad15c4e4a82e87408330).

File: packages/orchestration/src/exos/chain-hub.js [hunk 2]
```diff
@@ -216,6 +217,11 @@ export const makeChainHub = (zone, agoricNames, vowTools) => {
     keyShape: BrandShape,
     valueShape: M.string(),
   });
+  /** @type {MapStore<string, string>} */
+  const bech32PrefixToChainName = zone.mapStore('bech32PrefixToChainName', {
```
Comment (dckc) [inline packages/orchestration/src/exos/chain-hub.js:222]: The 1st arg to `makeScalarMapStore` is a tag for just the key, not the whole map, for use in error messages such as `bech32Prefix not found: blort`. So I was going to suggest:

```suggestion
  const bech32PrefixToChainName = zone.mapStore('bech32Prefix', {
```

but it looks like the arg here is a baggage key, so the longer name seems right.

File: packages/orchestration/src/exos/chain-hub.js [hunk 3]
```diff
@@ -425,6 +437,20 @@ export const makeChainHub = (zone, agoricNames, vowTools) => {
...
       return undefined;
     },
+    /**
+     * @param {string} address bech32 address
+     * @returns {CosmosChainInfo}
+     */
+    getChainInfoFromAddress(address) {
+      const [prefix, rest] = address.split('1');
+      (prefix && rest) ||
+        Fail`Address contains invalid bech32 separator: ${q(address)}`;
```
Comment (dckc) [inline packages/orchestration/src/exos/chain-hub.js]: cosmjs uses [bech32](https://www.npmjs.com/package/bech32), which uses `lastIndexOf`, not the 1st occurrence.

https://github.com/bitcoinjs/bech32/blob/5ceb0e3d4625561a459c85643ca6947739b2d83c/src/index.ts#L146

I wonder about factoring out a `getBech32Prefix(addr)` with its own unit tests.

File: packages/orchestration/src/exos/chain-hub.js [hunk 4]
```diff
@@ -425,6 +437,20 @@ export const makeChainHub = (zone, agoricNames, vowTools) => {
       }
       return undefined;
     },
+    /**
+     * @param {string} address bech32 address
+     * @returns {CosmosChainInfo}
+     */
+    getChainInfoFromAddress(address) {
```
Comment (dckc) [inline packages/orchestration/src/exos/chain-hub.js]: Did you consider `getChainInfoFromPrefix(getBech32Prefix(address))` instead or in addition?

File: packages/orchestration/src/exos/chain-hub.js (hunk 3 above)
Comment (dckc) [reply packages/orchestration/src/exos/chain-hub.js]: the so-called reference implementation uses `.rfind('1')` too.

https://github.com/sipa/bech32/blob/7a7d7ab158db7078a333384e0e918c90dbc42917/ref/python/segwit_addr.py#L79

**Review by dckc — CHANGES_REQUESTED** [review]
Strictly speaking, `lastIndexOf` is a correctness issue. I suppose it deserves a test.

factoring out `getBech32Prefix(addr)` is a suggestion; deal with it at your discretion.

File: packages/orchestration/src/exos/chain-hub.js (hunk 3 above)
Comment (0xpatrickdev) [reply packages/orchestration/src/exos/chain-hub.js]: Thanks  - glad I tagged you as a reviewer! Will factor this out to its own function and adopt the new logic.

File: packages/orchestration/src/exos/chain-hub.js (hunk 4 above)
Comment (0xpatrickdev) [reply packages/orchestration/src/exos/chain-hub.js]: I didn't, no.

> in addition

I think we should only have one function for this to reduce info overload.

> instead

I still feel `getChainInfoFromAddress` is the sensible abstraction, but happy to refactor to `getChainInfoFromPrefix` if you and other reviewers are in agreement that its preferable.

Comment (dckc) [reply packages/orchestration/src/exos/chain-hub.js]: This starts to look like a product question. `getChainInfoFromAddress` is what #10349 calls for, and I guess it got an OK from product.

I don't feel strongly about it. I guess `getChainInfoFromAddress` is fine by me.

File: packages/orchestration/src/exos/chain-hub.js (hunk 3 above)
Comment (0xpatrickdev) [reply packages/orchestration/src/exos/chain-hub.js]: Interestingly, in practice no one seems to use `1` in the prefix / human readable part:
https://github.com/satoshilabs/slips/blob/master/slip-0173.md#registered-human-readable-parts

But the spec **_does say_** any US-ASCII character, in the range [33-126], is valid for the HRP:
https://github.com/bitcoin/bips/blob/7a104491862379e88662a39d1a2b613c907e503e/bip-0173.mediawiki#bech32

Please see the amended commit: https://github.com/Agoric/agoric-sdk/pull/10354/commits/65e07f8db405c6c473ad62a64c2cb6f2166da159

File: packages/orchestration/src/exos/chain-hub.js (hunk 4 above)
Comment (0xpatrickdev) [reply packages/orchestration/src/exos/chain-hub.js]: For transparency, I wrote the ticket and prescribed the name there. I don't think product typically weighs in on function names, just the user story (here: "look up chainId from chain address value").

File: packages/orchestration/test/utils/address.test.ts [hunk 5]
```diff
@@ -107,3 +108,40 @@ test('makeICQChannelAddress', t => {
...
+      specimen: 'bc1qw508d6qejxtdg4y5r3zarvary0c5xw7kv8f3t4',
+      expected: 'bc',
+    },
+    {
+      specimen: 'cosmos1n4f2eqt2gm5mh6gevf8aw2wrf75q25yru09yvn',
+      expected: 'cosmos',
+    },
+    {
+      specimen: '111qw508d6qejxtdg4y5r3zarvary0c5xw7kv8f3t4',
+      expected: '11',
```
Comment (dckc) [inline packages/orchestration/test/utils/address.test.ts]: :clap: 

This test distinguishes the correct `.lastIndexOf()` implementation from the `indexOf()` style.

**Review by dckc — APPROVED** [review]

File: packages/orchestration/src/exos/chain-hub.js [hunk 6]
```diff
@@ -425,6 +438,18 @@ export const makeChainHub = (zone, agoricNames, vowTools) => {
...
+     * @param {string} address bech32 address
+     * @returns {CosmosChainInfo}
+     */
+    getChainInfoByAddress(address) {
+      const prefix = getBech32Prefix(address);
+      if (bech32PrefixToChainName.has(prefix)) {
+        const chainName = bech32PrefixToChainName.get(prefix);
+        return chainInfos.get(chainName);
+      }
+      throw makeError(`Chain info not found for bech32Prefix ${q(prefix)}`);
```
Comment (turadg) [inline packages/orchestration/src/exos/chain-hub.js]: consider: generally better to validate before getting to the happy path:
```suggestion
      if (!bech32PrefixToChainName.has(prefix)) {
        throw makeError(`Chain info not found for bech32Prefix ${q(prefix)}`);
      }

      const chainName = bech32PrefixToChainName.get(prefix);
      return chainInfos.get(chainName);
```

File: packages/orchestration/test/exos/chain-hub.test.ts [hunk 7]
```diff
@@ -177,3 +159,35 @@ test('toward asset info in agoricNames (#9572)', async t => {
     });
   }
 });
+
+test.serial('getChainInfoByAddress', async t => {
```
Comment (turadg) [inline packages/orchestration/test/exos/chain-hub.test.ts:163]: why is this serial now? each test has its own setup. 

we should avoid serial when we can

File: packages/orchestration/test/utils/address.test.ts [hunk 8]
```diff
@@ -107,3 +108,40 @@ test('makeICQChannelAddress', t => {
...
+      error:
+        'No separator character for "qw508d6qejxtdg4y5r3zarvary0c5xw7kv8f3t4"',
+    },
+    {
+      specimen: '1qw508d6qejxtdg4y5r3zarvary0c5xw7kv8f3t4',
+      error: 'Missing prefix for "1qw508d6qejxtdg4y5r3zarvary0c5xw7kv8f3t4"',
+    },
+  ];
+
+  for (const { specimen, expected, error } of testCases) {
```
Comment (turadg) [inline packages/orchestration/test/utils/address.test.ts]: consider an ava macro (e.g. `bech32`)  so you can have a test for each case:
```
test(bech32, 'bc1qw508d6qejxtdg4y5r3zarvary0c5xw7kv8f3t4', 'bc');
…
test(bech32, '1qw508d6qejxtdg4y5r3zarvary0c5xw7kv8f3t4', null, 'Missing prefix for "1qw508d6qejxtdg4y5r3zarvary0c5xw7kv8f3t4"');
```

File: packages/orchestration/test/exos/chain-hub.test.ts (hunk 7 above)
Comment (0xpatrickdev) [reply packages/orchestration/test/exos/chain-hub.test.ts:163]: Without this, and the addition of the `getChainInfoByAddress` test, `toward asset info in agoricNames (#9572)` starts failing with:

```
SES_UNHANDLED_REJECTION: (Error#1)
Error#1: promise watcher must be a virtual object
    at makeError (file:///Users/0xpatrick/repos/agoric-sdk/node_modules/ses/src/error/assert.js:350:61)
    at fail (file:///Users/0xpatrick/repos/agoric-sdk/node_modules/ses/src/error/assert.js:482:20)
    at Fail (file:///Users/0xpatrick/repos/agoric-sdk/node_modules/ses/src/error/assert.js:492:39)
    at file:///Users/0xpatrick/repos/agoric-sdk/packages/swingset-liveslots/src/watchedPromises.js:216:13
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)

  ─

  toward asset info in agoricNames (#9572)

  Error: Promise returned by test never resolved
```

Comment (turadg) [reply packages/orchestration/test/exos/chain-hub.test.ts:163]: Ok, out of scope. Turns out `provideDurableZone` mutates global state. I'll look into it

**Pull request merged**